Ideas/easier-embedded-rust.md

```diff
@@ -0,0 +1,57 @@
+# An Easier Embedded Rust
+
+## Problem Statement
+
+**How could Rust make it easier for people just getting started with Rust and/or Embedded?**
+
+Rust has made embedded development better (for me) in a number of ways:
+
+* Leveraging what Rust already gives you:
+    * Tools like Cargo for build + package management
+    * Ability to run `no_std` programs easily
+    * Modern language features (modules, iterators, type system)
+    * Making it "harder to get things wrong" - use the compiler to detect errors statically
+* Building good tools:
+    * Way better debugger tools via probe-rs, and tools built on top like probe-run and probe-rs-cli
+    * Tooling for logging, like defmt
+* Reasonably good docs, and teaching materials
+
+However, Rust does not have as good of a "getting started (new to embedded and/or new to rust" story, or a story for "just want things to work (artists/hobbyists)".
+
+For whatever shortcomings (performance, reliability, licensing, "under the hood complexity") exist (or are perceived to exist) for ecosystems like Arduino and CircuitPython/MicroPython: they make getting started much easier than Rust does today.
+
+## Concepts to consider
+
+Balancing "simple" and "easy". How can we make things *easier*, without hiding all of the complexity under the hood, or making things more complex just so we can hide them under the hood?
+
+Hidden complexity makes it very hard if you ever fall off the "happy path". Requiring "extra" tools not typically used by "production" embedded rust users makes it harder to find help. Special systems means effort is split.
+
+How can this be done efficiently? Arduino and CircuitPython both have corporate backing in addition to their open source efforts. While there is a growing open source embedded rust community, any new approaches would have to leverage existing work or start from scratch.
+
+How can this be done in a way that has a "path" to more traditional Rust usage? When users become more comfortable with embedded Rust, and would like to make lower level modifications, how can they do this without abandoning everything they've learned, or without experiencing a steep learning curve?
+
+## Potential Research Areas
+
+### Better enumeration of Why and How other approaches are easier
+
+It would be good to do an overview of today's Arduino and CircuitPython environments, and see what they provide (or don't) to make the experience easier. It would also be useful to capture where their approaches "fall over", or what limitations typical users run into.
```

Comment:
what do you mean by "fall over"?

Reply:
"fall over" is sort of a euphemism for failing, or when something stops working.

Sometimes you have a system that works, until it doesn't.

For example, if you had to fill out a form every time you shipped something, that might be fine for shipping one or two things a day. But if 100 orders come in one day, and you have to fill out a form by hand for all of those, it'd get overwhelming! This might be a point where the "filling out all forms by hand" approach "falls over", and you would want to switch to something like having a computer fill out the form for you.